### PR TITLE
DOC: Move Script Repository to readthedocs

### PR DESCRIPTION
**Pull request by lassoan**
This commit adds content of Slicer Script Repository (https://www.slicer.org/wiki/Documentation/Nightly/ScriptRepository) as of 2021-04-22 to Slicer documentation.

This makes the example scripts versioned together with the Slicer source code and makes it editable using standard github contribution process.

Restructured text format is used, as this is the preferred file format for developer documentation (more complicated syntax than markdown, but developers should be familiar with it anyway and it offers more structured organization and formatting).

Created by converting from mediawiki to rst using pandoc, then manually fixing formatting errors (mostly links, code blocks, and figures).

### CODE REVIEW

**Comment by lassoan** [comment]
Here is the generated preview of the page: https://slicer--5614.org.readthedocs.build/en/5614/developer_guide/script_repository.html

**Comment by sjh26** [comment]
👍

**Comment by jcfr** [comment]
Thanks for working on this :pray: 

Few comments:

* integration in table of content

| This PR | Suggested change |
|--|--|
| image | image |



* update references to `slice.util`

Since the python API associated with the `slice.util` module is generated, I suggest to use the following whenever possible

```
:func:`slicer.util.getNode()`
```

* Group all example in categories. It looks likes we could have move existing example into `Markups`, and also introduce sub categories like `Viewers`, `Loading/Saving`, ...

File: Docs/developer_guide/script_repository.rst [hunk 1]
```diff
@@ -0,0 +1,4237 @@
...
+
+Additional options may be specified in ``properties`` argument. For example, load an image stack by disabling ``singleFile`` option:
+
+.. code-block:: python
+
+   loadedVolumeNode = slicer.util.loadVolume('c:/Users/abc/Documents/SomeImage/file001.png', {'singleFile': False})
+
+Get a MRML node in the scene based on the node name and call methods of that object. For the MRHead sample data:
+
+.. code-block:: python
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:100]:
```suggestion
.. code-block:: python

```

Comment (jcfr) [reply Docs/developer_guide/script_repository.rst:100]: This add missing line

File: Docs/developer_guide/script_repository.rst [hunk 2]
```diff
@@ -0,0 +1,4237 @@
...
+   writer.SetFileName("c:/tmp/screenshot.png")
+   writer.SetInputConnection(wti.GetOutputPort())
+   writer.Write()
+
+Launching Slicer
+----------------
+
+-  How to open an .mrb file with Slicer at the command line?
+
+   Slicer.exe --python-code "slicer.util.loadScene( 'f:/2013-08-23-Scene.mrb' )"
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:79]:
```suggestion
::

   Slicer.exe --python-code "slicer.util.loadScene( 'f:/2013-08-23-Scene.mrb' )"
```

File: Docs/developer_guide/script_repository.rst [hunk 3]
```diff
@@ -0,0 +1,4237 @@
...
+
+-  How to run a script in the Slicer environment in batch mode (without showing any graphical user interface)?
+
+   Slicer.exe --python-code "doSomething; doSomethingElse; etc." --testing --no-splash --no-main-window
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:83]:
```suggestion
::

   Slicer.exe --python-code "doSomething; doSomethingElse; etc." --testing --no-splash --no-main-window
```

File: Docs/developer_guide/script_repository.rst [hunk 4]
```diff
@@ -0,0 +1,4237 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This code loads all DICOM objects into the scene from a file folder. All the registered plugins are evaluated and the one with the highest confidence will be used to load the data. Files are imported into a temporary DICOM database, so the current Slicer DICOM database is not impacted.
+
+.. code-block:: python
+
+   dicomDataDir = "c:/my/folder/with/dicom-files"  # input folder with DICOM files
+   loadedNodeIDs = []  # this list will contain the list of all loaded node IDs
+
+   from DICOMLib import DICOMUtils
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:229]: To suggest best practices, I am wondering if we should move the import at the top

File: Docs/developer_guide/script_repository.rst [hunk 5]
```diff
@@ -0,0 +1,4237 @@
...
+   incomingVolumeFolder = "c:/tmp/incoming"
+   incomingVolumesProcessed = []
+
+   def checkForNewVolumes():
+     # Check if there is a new file in the
+     from os import listdir
+     from os.path import isfile, join
+     for f in listdir(incomingVolumeFolder):
+       if f in incomingVolumesProcessed:
+         # this is an incoming file, it was already there
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:201]:
```suggestion
         # This is an incoming file, it was already there
```

File: Docs/developer_guide/script_repository.rst [hunk 6]
```diff
@@ -0,0 +1,4237 @@
...
+         continue
+       filePath = join(incomingVolumeFolder, f)
+       if not isfile(filePath):
+         # ignore directories
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:205]:
```suggestion
         # Ignore directories
```

Comment (jcfr) [reply Docs/developer_guide/script_repository.rst:205]: For consistency, I suggest to have the first letter of each comment capitalized

File: Docs/developer_guide/script_repository.rst [hunk 7]
```diff
@@ -0,0 +1,4237 @@
...
+   import pydicom
+   pydicom.dcmread(fileList[0])
+   ds.CTExposureSequence[0].ExposureModulationType
+
+How to access tag of a volume loaded from DICOM? For example, get the patient position stored in a volume:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+volumeName='2: ENT IMRT'
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:313]:
```suggestion
   volumeName = '2: ENT IMRT'
```

File: Docs/developer_guide/script_repository.rst [hunk 8]
```diff
@@ -0,0 +1,4237 @@
...
+     from os.path import isfile, join
+     for f in listdir(incomingVolumeFolder):
+       if f in incomingVolumesProcessed:
+         # this is an incoming file, it was already there
+         continue
+       filePath = join(incomingVolumeFolder, f)
+       if not isfile(filePath):
+         # ignore directories
+         continue
+       logging.info("Loading new file: "+f)
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:207]:
```suggestion
       logging.info("Loading new file: " + f)
```

Comment (jcfr) [reply Docs/developer_guide/script_repository.rst:207]: Add space around operators

File: Docs/developer_guide/script_repository.rst [hunk 9]
```diff
@@ -0,0 +1,4237 @@
...
+       accessToken="TfYXwbKAW7JYbAgZ7MyISf")
+
+How to access top level tags of DICOM images imported into Slicer?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For example, to print the first patient's first study's first series' "0020,0032" field:
+
+.. code-block:: python
+
+   db=slicer.dicomDatabase
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:278]:
```suggestion
   db = slicer.dicomDatabase
```

File: Docs/developer_guide/script_repository.rst [hunk 10]
```diff
@@ -0,0 +1,4237 @@
...
+   ds.CTExposureSequence[0].ExposureModulationType
+
+How to access tag of a volume loaded from DICOM? For example, get the patient position stored in a volume:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+volumeName='2: ENT IMRT'
+   n=slicer.util.getNode(volumeName)
+   instUids=n.GetAttribute('DICOM.instanceUIDs').split()
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:315]:
```suggestion
    instUids = n.GetAttribute("DICOM.instanceUIDs").split()
```

Comment (jcfr) [reply Docs/developer_guide/script_repository.rst:315]: In addition of adding spaces around the `=`, I also suggest we also consistently use double quotes

File: Docs/developer_guide/script_repository.rst [hunk 11]
```diff
@@ -0,0 +1,4237 @@
...
+
+   # example:
+   node=slicer.util.getNode('volume1')
+   path=self.pathFromNode(node)
+   print("DICOM path=%s" % path)
+
+How can I convert DICOM to NRRD on the command line?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+   /Applications/Slicer-4.6.2.app/Contents/MacOS/Slicer --no-main-window --python-code "node=slicer.util.loadVolume('/tmp/series/im0.dcm'); slicer.util.saveNode(node, '/tmp/output.nrrd'); exit()"
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:354]:
```suggestion
::

   /Applications/Slicer-4.6.2.app/Contents/MacOS/Slicer --no-main-window --python-code "node=slicer.util.loadVolume('/tmp/series/im0.dcm'); slicer.util.saveNode(node, '/tmp/output.nrrd'); exit()"
```

File: Docs/developer_guide/script_repository.rst [hunk 12]
```diff
@@ -0,0 +1,4237 @@
...
+
+-  How to turn on slice intersections in the crosshair menu on the toolbar:
+
+.. code-block:: python
+
+   viewNodes = slicer.util.getNodesByClass('vtkMRMLSliceCompositeNode')
+   for viewNode in viewNodes:
+     viewNode.SetSliceIntersectionVisibility(1)
+
+How to find similar functions? For this one I searched for "slice intersections" text in the whole slicer source code, found that the function is implemented in Base\QTGUI\qSlicerViewersToolBar.cxx, then translated the qSlicerViewersToolBarPrivate::setSliceIntersectionVisible(bool visible) method to Python.
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:489]:
```suggestion
  .. note::

     How to find similar functions?
     
     For this one I searched for "slice intersections" text in the whole slicer source code, found that the function is implemented in ``Base\QTGUI\qSlicerViewersToolBar.cxx``, then translated the ``qSlicerViewersToolBarPrivate::setSliceIntersectionVisible(bool visible)`` method to Python.
```

File: Docs/developer_guide/script_repository.rst [hunk 13]
```diff
@@ -0,0 +1,4237 @@
...
+
+How to define/edit a circular region of interest in a slice viewer?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drop two markup points on a slice view and copy-paste the code below into the Python console. After this, as you move the markups you’ll see a circle following the markups.
+
+.. code-block:: python
+
+   # Update the sphere from the fiducial points
+   def UpdateSphere(param1, param2):
```
Comment (jcfr) [inline Docs/developer_guide/script_repository.rst:520]:
```suggestion
   def UpdateSphere(param1, param2):
     """Update the sphere from the fiducial points
     """
```

Comment (jcfr) [reply Docs/developer_guide/script_repository.rst:520]: Use docstrings when it applies and remove comment

**Comment by lassoan** [comment]
@jcfr you can take over now, I've committed all your suggestions.

**Comment by jcfr** [comment]
There are still issue with some of heading not being rendered.

For example, in this section,  the `Get the pseudo-singleton subject hierarchy node` heading is not rendered:

image

image

**Comment by lassoan** [comment]
Headers are missing because of inconsistent underline character usage.

We should use these: https://www.odoo.com/documentation/user/14.0/contributing/documentation/rst_cheat_sheet.html#headings

@jcfr do you prefer to fix it or you would rather pass it back to me?

**Comment by jcfr** [comment]
I will let you take over the topic. I am done for the day. 🙏

**Comment by lassoan** [comment]
I've fixed all the issues, categorized all the snippets, split them up to multiple source files (still one rendered page). Please have a final look: https://slicer--5614.org.readthedocs.build/en/5614/developer_guide/script_repository.html

I've kept the "Script repository" as it contains a few shell scripts and we'll can move C++ examples there, too (sphinx has nice multi-language code blocks feature).

**Comment by jcfr** [comment]
The organization in categories looks great :100: 

Here is a suggestion to re-order them:
* Module specific categories are sorted alphabetically 

|  This PR |  Suggested |
|---|---|
| image | image|

**Comment by jcfr** [comment]
> I've kept the "Script repository" as it contains a few shell scripts and we'll can move C++ examples there, too (sphinx has nice multi-language code blocks feature).

Make sense

**Comment by jcfr** [comment]
I am also wondering if moving: `Slice views`, `3D views`, `Common view operations` and `Access VTK rendering classes` under a `Views` category would be sensible ?

**Pull request merged**